### PR TITLE
PHPC-1140: Implement Transactions specification

### PR DESCRIPTION
**Pull request by derickr**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -32,6 +32,9 @@ matrix:
     - php: 7.0
       env:
         - SERVER_VERSION=3.4.11
+    - php: 7.0
+      env:
+        - SERVER_VERSION=4.0.0-rc4
```
Comment (jmikola) [inline .travis.yml]: Relevant: https://github.com/mongodb/mongo-php-driver/pull/851

I'm not sure this will be helpful for the transaction tests (pending [PHPC-1184](https://jira.mongodb.org/browse/PHPC-1184)), since they require a replica set oplog to exist.

File: src/MongoDB/Manager.c [hunk 2]
```diff
@@ -669,11 +670,46 @@ static PHP_METHOD(Manager, startSession)
...
+		zval* txn_options = php_array_fetch_array(options, "defaultTransactionOptions");
+		mongoc_transaction_opt_t* txn_opts = NULL;
+
+		/* Thrown exception and return if the defaultTransactionOptions is not an array */
+		if (!txn_options) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"defaultTransactionOptions\" option to be an array");
+			if (cs_opts) {
+				mongoc_session_opts_destroy(cs_opts);
+			}
+			return;
```
Comment (jmikola) [inline src/MongoDB/Manager.c]: If we restructure the end of this function, can we utilize a `cleanup:` label? Consider:

```c
	cs = mongoc_client_start_session(intern->client, cs_opts, &error);

	if (cs) {
		phongo_session_init(return_value, cs TSRMLS_CC);
	} else {
		phongo_throw_exception_from_bson_error_t(&error TSRMLS_CC);
	}

cleanup:
	if (cs_opts) {
		mongoc_session_opts_destroy(cs_opts);
	}
}
```

Then we can use `goto cleanup;` here and within the `if (EG(exception))` block below.

File: src/MongoDB/Manager.c [hunk 3]
```diff
@@ -669,11 +670,46 @@ static PHP_METHOD(Manager, startSession)
...
 		mongoc_session_opts_set_causal_consistency(cs_opts, php_array_fetchc_bool(options, "causalConsistency"));
 	}
 
+	if (options && php_array_existsc(options, "defaultTransactionOptions")) {
+		zval* txn_options = php_array_fetch_array(options, "defaultTransactionOptions");
+		mongoc_transaction_opt_t* txn_opts = NULL;
+
+		/* Thrown exception and return if the defaultTransactionOptions is not an array */
+		if (!txn_options) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"defaultTransactionOptions\" option to be an array");
```
Comment (jmikola) [inline src/MongoDB/Manager.c]: For consistency with other unexpected type exceptions, can we change `php_array_fetch_array` above to `php_array_fetchc` and then use the following here?

```c
if (Z_TYPE_P(txn_options) != IS_ARRAY) {
	phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"defaultTransactionOptions\" option to be array, %s given", PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(txn_options));
}
```

File: src/MongoDB/Manager.c [hunk 4]
```diff
@@ -669,11 +670,46 @@ static PHP_METHOD(Manager, startSession)
...
 	}
 
-	if (options && php_array_exists(options, "causalConsistency")) {
+	if (options && php_array_existsc(options, "causalConsistency")) {
 		cs_opts = mongoc_session_opts_new();
 		mongoc_session_opts_set_causal_consistency(cs_opts, php_array_fetchc_bool(options, "causalConsistency"));
 	}
 
+	if (options && php_array_existsc(options, "defaultTransactionOptions")) {
+		zval* txn_options = php_array_fetch_array(options, "defaultTransactionOptions");
```
Comment (jmikola) [inline src/MongoDB/Manager.c]: Did you mean to use `php_array_fetchc_array` here? Though, see my comment below about restructuring before you address that.

File: src/MongoDB/Session.c [hunk 5]
```diff
@@ -237,6 +385,10 @@ ZEND_BEGIN_ARG_INFO_EX(ai_Session_advanceOperationTime, 0, 0, 1)
 	ZEND_ARG_INFO(0, timestamp)
 ZEND_END_ARG_INFO()
 
+ZEND_BEGIN_ARG_INFO_EX(ai_Session_startTransAction, 0, 0, 0)
```
Comment (jmikola) [inline src/MongoDB/Session.c]: I don't think the "A" should be capitalized: "startTransaction"

File: src/MongoDB/Session.c [hunk 6]
```diff
@@ -228,6 +230,152 @@ static PHP_METHOD(Session, getOperationTime)
...
+	SUPPRESS_UNUSED_WARNING(return_value_ptr)
+	SUPPRESS_UNUSED_WARNING(return_value_used)
+
+	intern = Z_SESSION_OBJ_P(getThis());
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		return;
+	}
+
+	if (!mongoc_client_session_commit_transaction(intern->client_session, NULL, &error)) {
```
Comment (jmikola) [inline src/MongoDB/Session.c]: Is there any value in capturing the `reply` to `commitTransaction` here?

[`txn_finish()` uses `mongoc_client_write_command_with_opts()`](https://github.com/mongodb/mongo-c-driver/blob/4f83296adbdec3159bccd4ef4e06a5d35be54bd4/src/libmongoc/src/mongoc/mongoc-client-session.c#L147) to execute the command, so we may want to check the error domain and conditionally upgrade a ServerException into a CommandException and stash the reply on it, as we do from [`executeCommand()` and friends](https://github.com/mongodb/mongo-php-driver/blob/4db394a36f496154e4c0a622df8716006aa12a5a/php_phongo.c#L930).

File: tests/session/session-startTransaction-001.phpt [hunk 7]
```diff
@@ -0,0 +1,23 @@
+--TEST--
+MongoDB\Driver\Session::startTransaction() mini test
+--SKIPIF--
+<?php require __DIR__ . "/../utils/basic-skipif.inc"; ?>
+<?php skip_if_server_version('<', '4.0'); ?>
```
Comment (jmikola) [inline tests/session/session-startTransaction-001.phpt]: AFAIK, transactions require a replica set oplog. You should probably add `skip_if_not_replica_set()` here as well.

All session-related stuff requires libmongoc crypto as well, so `skip_if_not_libmongoc_crypto()` should appear as the first "skip if" (for consistency). I believe all of the other session tests have that.

Should this

File: tests/session/session-startTransaction-001.phpt [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+require_once __DIR__ . "/../utils/basic.inc";
+
+$manager = new MongoDB\Driver\Manager(URI);
+$session = $manager->startSession();
+
+$session->startTransaction();
+$session->abortTransaction();
+
+$session->startTransaction();
+$session->commitTransaction();
```
Comment (jmikola) [inline tests/session/session-startTransaction-001.phpt]: I tested these methods against a 4.0.0-rc5 standalone and there were no errors in any of the transaction tests.

With regard to this test alone (I haven't reviewed the others yet), that may be due to the fact that a transaction is not actually started until the first command executed within in. See [`startTransaction`](https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#starttransaction) in the spec. Therefore, I would propose we do an insertOne or some other basic write within each transaction (between start and abort/commit).

Additionally, we may want another test to demonstrate an error when executing on a 4.0+ standalone (there are appropriate "skip_if" functions for that, too).

File: .travis.yml (hunk 1 above)
Comment (derickr) [reply .travis.yml]: Yeah, I am aware of that. But some tests just need a 4.0.0 one to run at, as they don't actually *talk* to the Database yet. But libmongoc still does the version check with a single server.

File: src/MongoDB/Session.c (hunk 5 above)
Comment (derickr) [reply src/MongoDB/Session.c]: Good catch.

File: tests/session/session-startTransaction-001.phpt (hunk 7 above)
Comment (derickr) [reply tests/session/session-startTransaction-001.phpt]: Happy to add the crypto stuff, but why do sessions require it?

File: tests/session/session-startTransaction-001.phpt (hunk 8 above)
Comment (derickr) [reply tests/session/session-startTransaction-001.phpt]: Replying to this and the above comment. This test does not require a working replicaset as it doesn't actually run a transaction. This test tests whether I could call the methods at all. libmongoc doesn't check whether there is a replica set (that's a server error), but it *does* test whether the server has the right wire protocol version (hence the ``< 4.0`` check).

File: src/MongoDB/Manager.c (hunk 2 above)
Comment (derickr) [reply src/MongoDB/Manager.c]: Yes, and a little more too.

File: src/MongoDB/Manager.c (hunk 3 above)
Comment (derickr) [reply src/MongoDB/Manager.c]: OK

File: src/MongoDB/Manager.c (hunk 4 above)
Comment (derickr) [reply src/MongoDB/Manager.c]: No longer relevant due to the previous comment.

File: src/MongoDB/Session.c (hunk 6 above)
Comment (derickr) [reply src/MongoDB/Session.c]: I think so, but as we're then doing this now in multiple places, we probably should create a helper method akin to ``phongo_throw_exception_from_bson_error_t`` — for example ``phongo_throw_exception_from_bson_error_and_reply`` ?

File: tests/session/session-startTransaction-001.phpt (hunk 8 above)
Comment (jmikola) [reply tests/session/session-startTransaction-001.phpt]: Noted. I'd suggest a comment to explain that. The title only said "mini test" so it wasn't clear what the intention was.

Even something like "ensure that methods can be called" as the title would add some more clarity.

File: src/MongoDB/Manager.c [hunk 9]
```diff
@@ -669,22 +671,53 @@ static PHP_METHOD(Manager, startSession)
...
 		phongo_session_init(return_value, cs TSRMLS_CC);
 	} else {
 		phongo_throw_exception_from_bson_error_t(&error TSRMLS_CC);
 	}
+
+cleanup:
+	if (cs_opts) {
+		mongoc_session_opts_destroy(cs_opts);
+	}
+	if (txn_opts) {
```
Comment (jmikola) [inline src/MongoDB/Manager.c]: Ah, were these never freed before? I didn't notice any memory leaks.

File: src/MongoDB/Session.c [hunk 10]
```diff
@@ -228,6 +230,153 @@ static PHP_METHOD(Session, getOperationTime)
...
+	SUPPRESS_UNUSED_WARNING(return_value_used)
+
+	intern = Z_SESSION_OBJ_P(getThis());
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		return;
+	}
+
+	if (!mongoc_client_session_commit_transaction(intern->client_session, &reply, &error)) {
+		phongo_throw_exception_from_bson_error_and_reply_t(&error, &reply TSRMLS_CC);
```
Comment (jmikola) [inline src/MongoDB/Session.c]: Test this for leaks. I think you still need a `bson_destroy(&reply);` after this.

`executeCommand()` does so before returning as its `free_reply` boolean will be true when it calls `phongo_throw_exception_from_bson_error_and_reply_t`.

File: scripts/presets/replicaset-30.json [hunk 11]
```diff
@@ -12,6 +12,7 @@
 				"noprealloc": true,
 				"nssize": 1,
 				"port": 3100,
+				"bind_ip_all": true,
```
Comment (jmikola) [inline scripts/presets/replicaset-30.json]: [`bind_ip_all` was introduced in 3.6](https://docs.mongodb.com/manual/reference/program/mongod/#cmdoption-mongod-bind-ip-all), so I don't think we should specify it for the configs using "30-release".

That said, I do think we should add it to all of the non-versioned configs, which will now be using "40-release". This includes all of the standalone JSON files apart from `standalone-30.json`.

File: scripts/ubuntu/mongo-orchestration-config.json [hunk 12]
```diff
@@ -1,5 +1,6 @@
 {
 	"releases": {
+		"40-release": "/home/vagrant/4.0/usr/bin",
 		"36-release": "/home/vagrant/3.6/usr/bin",
```
Comment (jmikola) [inline scripts/ubuntu/mongo-orchestration-config.json]: I don't think anything is using this. It was only important as the first version listed under "releases", which made it the default. Now that "40-release" is here, I think we can just remove this.

File: scripts/presets/replicaset-30.json (hunk 11 above)
Comment (jmikola) [reply scripts/presets/replicaset-30.json]: I suppose you can do `"bind_ip": "0.0.0.0,::"` for the 3.0 configs, per the documentation above.

File: tests/utils/tools.php [hunk 13]
```diff
@@ -45,7 +45,7 @@ function drop_collection($uri, $databaseName, $collectionName)
     $command = new Command(['drop' => $collectionName]);
 
     try {
-        $server->executeCommand($databaseName, $command);
+        $server->executeCommand($databaseName, $command, [ 'writeConcern' => new \MongoDB\Driver\WriteConcern( \MongoDB\Driver\WriteConcern::MAJORITY ) ] );
```
Comment (jmikola) [inline tests/utils/tools.php]: Why was this necessary? Did it interfere with tests that query the secondary and expected an empty collection?

File: tests/session/session-startTransaction_error-003.phpt [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+
+foreach ($options as $txnOptions) {
+    $session->startTransaction($txnOptions);
+}
+
+?>
+===DONE===
+<?php exit(0); ?>
+--EXPECTF--
+Warning: MongoDB\Driver\Session::startTransaction() expects parameter 1 to be array, integer given in %s%session-startTransaction_error-003.php on line %d
```
Comment (jmikola) [inline tests/session/session-startTransaction_error-003.phpt]: Other tests simply use `in %s on line %d`. Any reason not to do the same here and make the output a bit more robust?

File: scripts/presets/replicaset-30.json (hunk 11 above)
Comment (derickr) [reply scripts/presets/replicaset-30.json]: Done

File: tests/session/session-startTransaction_error-003.phpt (hunk 14 above)
Comment (derickr) [reply tests/session/session-startTransaction_error-003.phpt]: ah, crap. That's what you get doing it different in your other major (Xdebug) project. Will fix (again).

File: tests/utils/tools.php (hunk 13 above)
Comment (derickr) [reply tests/utils/tools.php]: No: https://jira.mongodb.org/browse/SERVER-35613 — I need to come up with whether I think this is Okay behaviour in the server though. It's certainly going to be mega confusing for users.

File: scripts/ubuntu/mongo-orchestration-config.json (hunk 12 above)
Comment (derickr) [reply scripts/ubuntu/mongo-orchestration-config.json]: We'll leave it for now.

File: src/MongoDB/Manager.c [hunk 15]
```diff
@@ -669,22 +671,53 @@ static PHP_METHOD(Manager, startSession)
...
 
-	cs = mongoc_client_start_session(intern->client, cs_opts, &error);
+	if (options && php_array_existsc(options, "defaultTransactionOptions")) {
+		zval* txn_options = php_array_fetchc(options, "defaultTransactionOptions");
 
-	if (cs_opts) {
-		mongoc_session_opts_destroy(cs_opts);
+		/* Thrown exception and return if the defaultTransactionOptions is not an array */
+		if (Z_TYPE_P(txn_options) != IS_ARRAY) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"defaultTransactionOptions\" option to be an array");
```
Comment (jmikola) [inline src/MongoDB/Manager.c]: I think my original request in https://github.com/mongodb/mongo-php-driver/pull/850#discussion_r195877062 was lost by subsequent edits. Can we add ", %s given" for consistency with other exceptions? The inconsistency stands out in the error tests you have below, which is also how I realized this change wasn't made.

File: src/MongoDB/Session.c [hunk 16]
```diff
@@ -228,6 +230,154 @@ static PHP_METHOD(Session, getOperationTime)
...
+
+		mongoc_transaction_opts_set_read_prefs(opts, phongo_read_preference_from_zval(read_preference TSRMLS_CC));
+	}
+
+	if (php_array_existsc(options, "writeConcern")) {
+		zval* write_concern = php_array_fetchc(options, "writeConcern");
+
+		if (Z_TYPE_P(write_concern) != IS_OBJECT || !instanceof_function(Z_OBJCE_P(write_concern), php_phongo_writeconcern_ce TSRMLS_CC)) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"writeConcern\" option to be %s, %s given", ZSTR_VAL(php_phongo_writeconcern_ce->name), PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(write_concern));
+			return false;
```
Comment (jmikola) [inline src/MongoDB/Session.c]: Technically, it may be possible for this to leak if a valid ReadConcern or ReadPreference is added to `mongoc_transaction_opt_t* opts` but the WriteConcern fails. Likewise for a valid ReadConcern and invalid ReadPreference. Both of these cases should be easy to test and demonstrate a leak. I imagine the current error tests didn't expose this because the test cases each contain one option at a time.

Should just require adding the following here and in ReadPreference:

```c
if (opts) {
	mongoc_transaction_opts_destroy(opts);
}
```

File: tests/session/session-startTransaction_error-002.phpt [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+$options = [
+    [ 'readConcern' => 42 ], 
+    [ 'readConcern' => new stdClass ],
+    [ 'readConcern' => new \MongoDB\Driver\WriteConcern( 2 ) ],
+    [ 'readPreference' => 42 ], 
+    [ 'readPreference' => new stdClass ],
+    [ 'readPreference' => new \MongoDB\Driver\ReadConcern( \MongoDB\Driver\ReadConcern::LOCAL ) ],
+    [ 'writeConcern' => 42 ], 
+    [ 'writeConcern' => new stdClass ],
+    [ 'writeConcern' => new \MongoDB\Driver\ReadPreference( \MongoDB\Driver\ReadPreference::RP_SECONDARY ) ],
```
Comment (jmikola) [inline tests/session/session-startTransaction_error-002.phpt]: I'd suggest two or three new cases:

 * Valid ReadConcern and invalid ReadPreference
 * Valid ReadConcern and invalid WriteConcern (optional)	
 * Valid ReadPreference and invalid WriteConcern	

This would test the memory leak I noted above.

File: tests/session/session_error-001.phpt [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+$options = [
+    [ 'readConcern' => 42 ], 
+    [ 'readConcern' => new stdClass ],
+    [ 'readConcern' => new \MongoDB\Driver\WriteConcern( 2 ) ],
+    [ 'readPreference' => 42 ], 
+    [ 'readPreference' => new stdClass ],
+    [ 'readPreference' => new \MongoDB\Driver\ReadConcern( \MongoDB\Driver\ReadConcern::LOCAL ) ],
+    [ 'writeConcern' => 42 ], 
+    [ 'writeConcern' => new stdClass ],
+    [ 'writeConcern' => new \MongoDB\Driver\ReadPreference( \MongoDB\Driver\ReadPreference::RP_SECONDARY ) ],
```
Comment (jmikola) [inline tests/session/session_error-001.phpt]: Likewise, this can include test cases I suggested in https://github.com/mongodb/mongo-php-driver/pull/850/files#r196199595 (redundant, but not harm in keeping the tests consistent).

File: tests/session/transaction-integration-001.phpt [hunk 19]
```diff
@@ -0,0 +1,82 @@
+--TEST--
+MongoDB\Driver\Session::startTransaction() mini test
```
Comment (jmikola) [inline tests/session/transaction-integration-001.phpt]: Would "Committing a transaction" be a better name? I assume this doesn't encounter any errors and just demonstrates a transaction across two collections.

File: tests/session/transaction-integration-002.phpt [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+<?php skip_if_not_clean(DATABASE_NAME, COLLECTION_NAME); ?>
+<?php skip_if_not_replica_set(); ?>
+<?php skip_if_server_version('<', '4.0'); ?>
+--FILE--
+<?php
+require_once __DIR__ . "/../utils/basic.inc";
+
+$manager = new MongoDB\Driver\Manager(URI);
+
+/* Create collections as that can't be (automatically) done in a transaction */
```
Comment (jmikola) [inline tests/session/transaction-integration-002.phpt]: So no implicit creation of collections/databases within a transaction? I wasn't aware of that limitation.

Comment (jmikola) [reply tests/session/transaction-integration-002.phpt]: I suppose the `create` command doesn't have the same issue with `drop` in that it requires a majority WC, correct?

File: src/MongoDB/Manager.c (hunk 9 above)
Comment (jmikola) [reply src/MongoDB/Manager.c]: I think https://github.com/mongodb/mongo-php-driver/pull/850/files#r196198256 highlights a possible leak that I may have been thinking off, since that's one case where `php_mongodb_session_parse_transaction_options()` could return 0 while still having allocated memory.

Once that leak is addressed, I don't think we need this in `cleanup:` as we can call `mongoc_transaction_opts_destroy()` immediately after setting it on the session options.

File: src/MongoDB/Manager.c [hunk 21]
```diff
@@ -669,22 +671,53 @@ static PHP_METHOD(Manager, startSession)
...
+			goto cleanup;
+		}
+
+		/* If the options are non-empty, add them to the client session opts struct */
+		if (txn_opts) {
+			if (!cs_opts) {
+				cs_opts = mongoc_session_opts_new();
+			}
+
+			mongoc_session_opts_set_default_transaction_opts(cs_opts, txn_opts);
```
Comment (jmikola) [inline src/MongoDB/Manager.c]: Can we call `mongoc_transaction_opts_destroy()` after this? Ideally, this should be the only point where `txn_opts` is non-NULL and in need of freeing (provided the memory leak in https://github.com/mongodb/mongo-php-driver/pull/850/files#r196198256 is fixed).

File: src/MongoDB/Manager.c (hunk 15 above)
Comment (derickr) [reply src/MongoDB/Manager.c]: Done

File: src/MongoDB/Session.c [hunk 22]
```diff
@@ -228,6 +230,154 @@ static PHP_METHOD(Session, getOperationTime)
...
+mongoc_transaction_opt_t* php_mongodb_session_parse_transaction_options(zval* options TSRMLS_DC)
+{
+	mongoc_transaction_opt_t* opts = NULL;
+
+	if (php_array_existsc(options, "readConcern")) {
+		zval* read_concern = php_array_fetchc(options, "readConcern");
+
+		if (Z_TYPE_P(read_concern) != IS_OBJECT || !instanceof_function(Z_OBJCE_P(read_concern), php_phongo_readconcern_ce TSRMLS_CC)) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"readConcern\" option to be %s, %s given", ZSTR_VAL(php_phongo_readconcern_ce->name), PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(read_concern));
+			return false;
```
[the remark this answers is not in the document]
Comment (derickr) [reply src/MongoDB/Session.c]: Done

File: src/MongoDB/Session.c (hunk 16 above)
Comment (derickr) [reply src/MongoDB/Session.c]: Done

File: tests/session/session-startTransaction_error-002.phpt (hunk 17 above)
Comment (derickr) [reply tests/session/session-startTransaction_error-002.phpt]: Done

File: tests/session/session_error-001.phpt (hunk 18 above)
Comment (derickr) [reply tests/session/session_error-001.phpt]: Done

File: tests/session/transaction-integration-002.phpt (hunk 20 above)
Comment (derickr) [reply tests/session/transaction-integration-002.phpt]: Correct. ``create`` isn't two phase, I believe.

And yes, just like older MySQL's, support for DDL changes in transactions doesn't seem to be included. At least we get an exception instead of silently ignoring it.

File: tests/session/transaction-integration-001.phpt (hunk 19 above)
Comment (derickr) [reply tests/session/transaction-integration-001.phpt]: I've done that. It's basically the same as the example from the PHPLIB requested docs: https://jira.mongodb.org/browse/PHPLIB-350

File: src/MongoDB/Manager.c (hunk 21 above)
Comment (derickr) [reply src/MongoDB/Manager.c]: Yes, done. And updated relevant comments.

File: src/MongoDB/Session.c [hunk 23]
```diff
@@ -228,6 +230,166 @@ static PHP_METHOD(Session, getOperationTime)
...
+	mongoc_transaction_opt_t* opts = NULL;
+
+	if (php_array_existsc(options, "readConcern")) {
+		zval* read_concern = php_array_fetchc(options, "readConcern");
+
+		if (Z_TYPE_P(read_concern) != IS_OBJECT || !instanceof_function(Z_OBJCE_P(read_concern), php_phongo_readconcern_ce TSRMLS_CC)) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"readConcern\" option to be %s, %s given", ZSTR_VAL(php_phongo_readconcern_ce->name), PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(read_concern));
+			/* Freeing opts is not needed here, as it can't be set yet. The
+			 * code is here to keep it consistent with the others in case more
+			 * options are added before this one. */
```
Comment (jmikola) [inline src/MongoDB/Session.c]: Thanks for the comment :+1:

File: tests/session/transaction-integration-001.phpt [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+    }
+}
+
+?>
+===DONE===
+<?php exit(0); ?>
+--EXPECTF--
+Transaction committed.
+===DONE===
+
```
Comment (jmikola) [inline tests/session/transaction-integration-001.phpt]: Extra line break here?

File: tests/session/transaction-integration-002.phpt [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+    echo (isset($rd->errorLabels) && in_array("TransientTransactionError", $rd->errorLabels)) ?
+        "found a TransientTransactionError" : "did NOT get a TransientTransactionError", "\n";
+}
+?>
+===DONE===
+<?php exit(0); ?>
+--EXPECTF--
+found a TransientTransactionError
+===DONE===
+
```
Comment (jmikola) [inline tests/session/transaction-integration-002.phpt]: Extra line break here?

File: tests/utils/tools.php (hunk 13 above)
Comment (jmikola) [reply tests/utils/tools.php]: Reminder to add a comment and reference to SERVER-35613.
